From 8c6a5d83c51f5c4727bc3286ebefd41926f41e17 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sat, 13 Dec 2025 22:32:05 -0600
Subject: [PATCH 413/479] atidxx: Fixup vtable layouts.

finally the mystery is solved
---
 dlls/atidxx64/atidxx.h | 14 ++++---
 dlls/atidxx64/main.c   | 86 ++++++++++++++++++++++++------------------
 2 files changed, 58 insertions(+), 42 deletions(-)

diff --git a/dlls/atidxx64/atidxx.h b/dlls/atidxx64/atidxx.h
index b57776daafd..9a152206919 100644
--- a/dlls/atidxx64/atidxx.h
+++ b/dlls/atidxx64/atidxx.h
@@ -48,6 +48,8 @@ DECLARE_INTERFACE(IAmdDxExtInterface)
 {
     THISCALLMETHOD_(unsigned int, AddRef)(THIS) PURE;
     THISCALLMETHOD_(unsigned int, Release)(THIS) PURE;
+    /* virtual destructor, somehow missed this originally... im so dumb */
+    THISCALLMETHOD_(void, Destroy)(THIS) PURE;
 };
 #undef INTERFACE
 
@@ -57,18 +59,19 @@ DECLARE_INTERFACE_(IAmdDxExt, IAmdDxExtInterface)
     /*** IAmdDxExtInterface methods ***/
     THISCALLMETHOD_(unsigned int, AddRef)(THIS) PURE;
     THISCALLMETHOD_(unsigned int, Release)(THIS) PURE;
+    THISCALLMETHOD_(void, Destroy)(THIS) PURE;
 
     /*** IAmdDxExt methods ***/
-    THISCALLMETHOD_(HRESULT, IaGetPrimitiveTopology)(THIS_ AmdDxExtPrimitiveTopology *topology) PURE;
     THISCALLMETHOD_(HRESULT, GetVersion)(THIS_ AmdDxExtVersion *version) PURE;
     THISCALLMETHOD_(IAmdDxExtInterface*,GetExtInterface)(THIS_ unsigned int iface) PURE;
 
 
     THISCALLMETHOD_(HRESULT, IaSetPrimitiveTopology)(THIS_ D3D_PRIMITIVE_TOPOLOGY topology) PURE;
+    THISCALLMETHOD_(HRESULT, IaGetPrimitiveTopology)(THIS_ AmdDxExtPrimitiveTopology *topology) PURE;
     THISCALLMETHOD_(HRESULT, SetSingleSampleRead)(THIS_ ID3D10Resource *res, BOOL single_sample) PURE;
     THISCALLMETHOD_(HRESULT, SetSingleSampleRead11)(THIS_ ID3D11Resource *res, BOOL single_sample) PURE;
-    THISCALLMETHOD_(HRESULT, IaSetPrimitiveTopologyCtx)(THIS_ unsigned int topology, ID3D11DeviceContext *ctx) PURE;
     THISCALLMETHOD_(HRESULT, QueryFeatureSupport)(THIS_ unsigned int feature_token, void *data, unsigned int data_size) PURE;
+    THISCALLMETHOD_(HRESULT, IaSetPrimitiveTopologyCtx)(THIS_ unsigned int topology, ID3D11DeviceContext *ctx) PURE;
     THISCALLMETHOD_(HRESULT, IaGetPrimitiveTopologyCtx)(THIS_ AmdDxExtPrimitiveTopology *topology, ID3D11DeviceContext *ctx) PURE;
 };
 #undef INTERFACE
@@ -79,9 +82,9 @@ DECLARE_INTERFACE_(IAmdDxExtUAVOverlap, IAmdDxExtInterface)
     /*** IAmdDxExtInterface methods ***/
     THISCALLMETHOD_(unsigned int, AddRef)(THIS) PURE;
     THISCALLMETHOD_(unsigned int, Release)(THIS) PURE;
+    THISCALLMETHOD_(void, Destroy)(THIS) PURE;
 
     /*** IAmdDxExtUAVOverlap methods ***/
-    THISCALLMETHOD_(void, Unk1)(THIS) PURE;
     THISCALLMETHOD_(HRESULT, BeginUAVOverlap)(THIS) PURE;
     THISCALLMETHOD_(HRESULT, EndUAVOverlap)(THIS) PURE;
     THISCALLMETHOD_(void, GetVersion)(THIS, AmdDxExtVersion* version) PURE;
@@ -94,6 +97,7 @@ DECLARE_INTERFACE_(IAmdDxExtQuadBufferStereo, IAmdDxExtInterface)
     /*** IAmdDxExtInterface methods ***/
     THISCALLMETHOD_(unsigned int, AddRef)(THIS) PURE;
     THISCALLMETHOD_(unsigned int, Release)(THIS) PURE;
+    THISCALLMETHOD_(void, Destroy)(THIS) PURE;
 
     /*** IAmdDxExtQuadBufferStereo methods ***/
     THISCALLMETHOD_(HRESULT, EnableQuadBufferStereo)(THIS, BOOL enable) PURE;
@@ -108,9 +112,9 @@ DECLARE_INTERFACE_(IAmdDxExtDepthBounds, IAmdDxExtInterface)
     /*** IAmdDxExtInterface methods ***/
     THISCALLMETHOD_(unsigned int, AddRef)(THIS) PURE;
     THISCALLMETHOD_(unsigned int, Release)(THIS) PURE;
+    THISCALLMETHOD_(void, Destroy)(THIS) PURE;
 
     /*** IAmdDxExtDepthBounds methods ***/
-    THISCALLMETHOD_(void, Unk1)(THIS) PURE;
     THISCALLMETHOD_(HRESULT, SetDepthBounds)(THIS, BOOL enabled, float min, float max) PURE;
     THISCALLMETHOD_(void, GetVersion)(THIS, AmdDxExtVersion *version) PURE;
 };
@@ -122,9 +126,9 @@ DECLARE_INTERFACE_(IAmdDxExtMultidrawIndirect, IAmdDxExtInterface)
     /*** IAmdDxExtInterface methods ***/
     THISCALLMETHOD_(unsigned int, AddRef)(THIS) PURE;
     THISCALLMETHOD_(unsigned int, Release)(THIS) PURE;
+    THISCALLMETHOD_(void, Destroy)(THIS) PURE;
 
     /*** IAmdDxExtMultidrawIndirect methods ***/
-    THISCALLMETHOD_(void, Unk1)(THIS) PURE;
     THISCALLMETHOD_(void, GetVersion)(THIS, AmdDxExtVersion *version) PURE;
     THISCALLMETHOD_(HRESULT, MultiDrawIndirect)(THIS, unsigned int draw_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride) PURE;
     THISCALLMETHOD_(HRESULT, MultiDrawIndexedIndirect)(THIS, unsigned int draw_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride) PURE;
diff --git a/dlls/atidxx64/main.c b/dlls/atidxx64/main.c
index 2f8355cb34f..de38ad36a6c 100644
--- a/dlls/atidxx64/main.c
+++ b/dlls/atidxx64/main.c
@@ -100,7 +100,7 @@ HRESULT WINAPI AmdD3D11CreateDeviceExt(IDXGIAdapter *adapter, D3D_DRIVER_TYPE dr
 
     0x0 = AddRef
     0x8 = Release
-    0x10 = ??
+    0x10 = virtual destructor
     0x18 = BeginUAVOverlap
     0x20 = EndUAVOverlap
     0x28 = GetVersion (called on init, prob some kind of version getter)
@@ -110,7 +110,7 @@ HRESULT WINAPI AmdD3D11CreateDeviceExt(IDXGIAdapter *adapter, D3D_DRIVER_TYPE dr
 /* field_0x168 (0xb)
     0x0 = AddRef
     0x8 = Release
-    0x10 = ??
+    0x10 = virtual destructor
     0x18 = SetDepthBounds
     0x20 = GetVersion
 */
@@ -118,7 +118,7 @@ HRESULT WINAPI AmdD3D11CreateDeviceExt(IDXGIAdapter *adapter, D3D_DRIVER_TYPE dr
 /* field_0x170 (0x11)
     0x0 = AddRef
     0x8 = Release
-    0x10 = ??
+    0x10 = virtual destructor
     0x18 = GetVersion
     0x20 = MultiDrawIndirect
     0x28 = MultiDrawIndexedIndirect
@@ -195,6 +195,12 @@ unsigned int __thiscall AmdDxExt_Release(IAmdDxExt *iface)
     return ref;
 }
 
+DEFINE_THISCALL_WRAPPER(AmdDxExt_Release, 4)
+void __thiscall AmdDxExt_Destroy(IAmdDxExt *iface)
+{
+    FIXME("%p stub!\n", iface);
+}
+
 DEFINE_THISCALL_WRAPPER(AmdDxExt_GetVersion, 8)
 HRESULT __thiscall AmdDxExt_GetVersion(IAmdDxExt *ext, AmdDxExtVersion *version)
 {
@@ -206,20 +212,6 @@ HRESULT __thiscall AmdDxExt_GetVersion(IAmdDxExt *ext, AmdDxExtVersion *version)
     return S_OK;
 }
 
-DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_AddRef, 4)
-unsigned int __thiscall AmdDxExtUAVOverlap_AddRef(IAmdDxExtUAVOverlap *iface)
-{
-    AmdDxExt *This = impl_from_IAmdDxExtUAVOverlap(iface);
-    return AmdDxExt_AddRef(&This->IAmdDxExt_iface);
-}
-
-DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_Release, 4)
-unsigned int __thiscall AmdDxExtUAVOverlap_Release(IAmdDxExtUAVOverlap *iface)
-{
-    AmdDxExt *This = impl_from_IAmdDxExtUAVOverlap(iface);
-    return AmdDxExt_Release(&This->IAmdDxExt_iface);
-}
-
 DEFINE_THISCALL_WRAPPER(AmdDxExt_GetExtInterface, 8)
 IAmdDxExtInterface* __thiscall AmdDxExt_GetExtInterface(IAmdDxExt *ext, unsigned int iface)
 {
@@ -248,10 +240,7 @@ IAmdDxExtInterface* __thiscall AmdDxExt_GetExtInterface(IAmdDxExt *ext, unsigned
         }
     }
 
-    if (ret)
-    {
-        AmdDxExt_AddRef(ext);
-    }
+    if (ret) AmdDxExt_AddRef(ext);
 
     return ret;
 }
@@ -318,10 +307,25 @@ HRESULT __thiscall AmdDxExt_IaGetPrimitiveTopologyCtx(IAmdDxExt *iface, AmdDxExt
     return E_NOTIMPL;
 }
 
-DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_Unk1, 4)
-void __thiscall AmdDxExtUAVOverlap_Unk1(IAmdDxExtUAVOverlap *iface)
+/* our release method handles destruction */
+DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_Destroy, 4)
+void __thiscall AmdDxExtUAVOverlap_Destroy(IAmdDxExtUAVOverlap *iface)
 {
-    FIXME("%p stub\n", iface);
+    FIXME("%p stub!\n", iface);
+}
+
+DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_AddRef, 4)
+unsigned int __thiscall AmdDxExtUAVOverlap_AddRef(IAmdDxExtUAVOverlap *iface)
+{
+    AmdDxExt *This = impl_from_IAmdDxExtUAVOverlap(iface);
+    return AmdDxExt_AddRef(&This->IAmdDxExt_iface);
+}
+
+DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_Release, 4)
+unsigned int __thiscall AmdDxExtUAVOverlap_Release(IAmdDxExtUAVOverlap *iface)
+{
+    AmdDxExt *This = impl_from_IAmdDxExtUAVOverlap(iface);
+    return AmdDxExt_Release(&This->IAmdDxExt_iface);
 }
 
 DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_BeginUAVOverlap, 4)
@@ -376,6 +380,12 @@ unsigned int __thiscall AmdDxExtQuadBufferStereo_Release(IAmdDxExtQuadBufferSter
     return AmdDxExt_Release(&This->IAmdDxExt_iface);
 }
 
+DEFINE_THISCALL_WRAPPER(AmdDxExtQuadBufferStereo_Destroy, 4)
+void __thiscall AmdDxExtQuadBufferStereo_Destroy(IAmdDxExtQuadBufferStereo *iface)
+{
+    FIXME("%p stub!\n", iface);
+}
+
 DEFINE_THISCALL_WRAPPER(AmdDxExtQuadBufferStereo_EnableQuadBufferStereo, 8)
 HRESULT __thiscall AmdDxExtQuadBufferStereo_EnableQuadBufferStereo(IAmdDxExtQuadBufferStereo *iface, BOOL enable)
 {
@@ -442,8 +452,8 @@ HRESULT __thiscall AmdDxExtDepthBounds_SetDepthBounds(IAmdDxExtDepthBounds *ifac
     return S_OK;
 }
 
-DEFINE_THISCALL_WRAPPER(AmdDxExtDepthBounds_Unk1, 4)
-void __thiscall AmdDxExtDepthBounds_Unk1(IAmdDxExtDepthBounds *iface)
+DEFINE_THISCALL_WRAPPER(AmdDxExtDepthBounds_Destroy, 4)
+void __thiscall AmdDxExtDepthBounds_Destroy(IAmdDxExtDepthBounds *iface)
 {
     FIXME("%p stub\n", iface);
 }
@@ -462,8 +472,8 @@ unsigned int __thiscall AmdDxExtMultidrawIndirect_Release(IAmdDxExtMultidrawIndi
     return AmdDxExt_Release(&This->IAmdDxExt_iface);
 }
 
-DEFINE_THISCALL_WRAPPER(AmdDxExtMultidrawIndirect_Unk1, 4)
-void __thiscall AmdDxExtMultidrawIndirect_Unk1(IAmdDxExtMultidrawIndirect *iface)
+DEFINE_THISCALL_WRAPPER(AmdDxExtMultidrawIndirect_Destroy, 4)
+void __thiscall AmdDxExtMultidrawIndirect_Destroy(IAmdDxExtMultidrawIndirect *iface)
 {
     FIXME("%p stub\n", iface);
 }
@@ -571,7 +581,7 @@ static const IAmdDxExtUAVOverlapVtbl amddxext_uav_vtable =
 {
     THISCALL(AmdDxExtUAVOverlap_AddRef),
     THISCALL(AmdDxExtUAVOverlap_Release),
-    THISCALL(AmdDxExtUAVOverlap_Unk1),
+    THISCALL(AmdDxExtUAVOverlap_Destroy),
     THISCALL(AmdDxExtUAVOverlap_BeginUAVOverlap),
     THISCALL(AmdDxExtUAVOverlap_EndUAVOverlap),
     THISCALL(AmdDxExtUAVOverlap_GetVersion)
@@ -581,21 +591,23 @@ static const IAmdDxExtVtbl AmdDxExt_vtable =
 {
     THISCALL(AmdDxExt_AddRef), //0
     THISCALL(AmdDxExt_Release), //0x8
-    THISCALL(AmdDxExt_IaGetPrimitiveTopology), // 0x10 ??
+    THISCALL(AmdDxExt_Destroy), //0x10
     THISCALL(AmdDxExt_GetVersion), //0x18
     THISCALL(AmdDxExt_GetExtInterface), //0x20
     THISCALL(AmdDxExt_IaSetPrimitiveTopology), //0x28
-    THISCALL(AmdDxExt_SetSingleSampleRead), //0x30 ??
-    THISCALL(AmdDxExt_SetSingleSampleRead11), //0x38 ??
-    THISCALL(AmdDxExt_IaSetPrimitiveTopologyCtx), //0x40 ??
+    THISCALL(AmdDxExt_IaGetPrimitiveTopology), // 0x30
+    THISCALL(AmdDxExt_SetSingleSampleRead), //0x38 ??
+    THISCALL(AmdDxExt_SetSingleSampleRead11), //0x40 ??
     THISCALL(AmdDxExt_QueryFeatureSupport), //0x48
-    THISCALL(AmdDxExt_IaGetPrimitiveTopologyCtx) //0x50 ??
+    THISCALL(AmdDxExt_IaSetPrimitiveTopologyCtx), //0x50 ??
+    THISCALL(AmdDxExt_IaGetPrimitiveTopologyCtx) //0x58 ??
 };
 
 static const IAmdDxExtQuadBufferStereoVtbl quadbufstereo_vtable =
 {
     THISCALL(AmdDxExtQuadBufferStereo_AddRef),
     THISCALL(AmdDxExtQuadBufferStereo_Release),
+    THISCALL(AmdDxExtQuadBufferStereo_Destroy),
     THISCALL(AmdDxExtQuadBufferStereo_EnableQuadBufferStereo),
     THISCALL(AmdDxExtQuadBufferStereo_GetLineOffset),
     THISCALL(AmdDxExtQuadBufferStereo_GetDisplayModeList),
@@ -605,7 +617,7 @@ static const IAmdDxExtDepthBoundsVtbl amddxext_depth_vtable =
 {
     THISCALL(AmdDxExtDepthBounds_AddRef),
     THISCALL(AmdDxExtDepthBounds_Release),
-    THISCALL(AmdDxExtDepthBounds_Unk1),
+    THISCALL(AmdDxExtDepthBounds_Destroy),
     THISCALL(AmdDxExtDepthBounds_SetDepthBounds),
     THISCALL(AmdDxExtDepthBounds_GetVersion)
 };
@@ -614,7 +626,7 @@ static const IAmdDxExtMultidrawIndirectVtbl amddxext_multidraw_vtable =
 {
     THISCALL(AmdDxExtMultidrawIndirect_AddRef),
     THISCALL(AmdDxExtMultidrawIndirect_Release),
-    THISCALL(AmdDxExtMultidrawIndirect_Unk1),
+    THISCALL(AmdDxExtMultidrawIndirect_Destroy),
     THISCALL(AmdDxExtMultidrawIndirect_GetVersion),
     THISCALL(AmdDxExtMultidrawIndirect_MultiDrawIndirect),
     THISCALL(AmdDxExtMultidrawIndirect_MultiDrawIndexedIndirect),
@@ -624,10 +636,10 @@ static const IAmdDxExtMultidrawIndirectVtbl amddxext_multidraw_vtable =
 
 HRESULT CDECL AmdDxExtCreate11(ID3D11Device *device, IAmdDxExt **ext)
 {
+    UINT64 id;
     HRESULT ret;
     AmdDxExt *obj;
     ID3D11VkExtDevice *ext_device;
-    UINT64 id;
     TRACE("%p %p\n", device, ext);
 
     if (!ext) return E_INVALIDARG;
-- 
2.52.0

